### PR TITLE
MARKET-1641 Is there a way to make popover position dynamic?

### PR DESCRIPTION
**Pull request by StasivPavlo**

### CODE REVIEW

File: components/bl-popover-component/src/Tooltip.js [hunk 1]
```diff
@@ -1,7 +1,8 @@
-import { useEffect, useMemo, useCallback } from 'react';
+import { useEffect, useMemo, useCallback, useState } from 'react';
 import { translatePopover } from './helpers';
 
-export function Tooltip({ contentElement, position, popoverContent }) {
+export function Tooltip({ targetRef, position, popoverContent }) {
+  const [validatePosition, setValidatePosition] = useState(position);
```
Comment (v-excelsior) [inline components/bl-popover-component/src/Tooltip.js:5]: validate - it's an action, I believe there should be valid or validated

File: components/bl-popover-component/src/helpers.js [hunk 2]
```diff
@@ -1,45 +1,72 @@
...
     [Position.BOTTOM]: {
-      topShift : contentOffset.top + contentHeight - rootOffset.top + root.firstChild.clientHeight,
+      topShift : targetOffset.top + targetHeight - contentOffset.top + contentElement.firstChild.clientHeight,
       leftShift: verticalLeftShift
     }
   };
 
-  return ShiftHandler[position];
+  const PositionCheck = {
+    left  : () => (
```
Comment (v-excelsior) [inline components/bl-popover-component/src/helpers.js:47]: left => Position.LEFT - should be same as for ShiftHandler

File: components/bl-popover-component/src/helpers.js [hunk 3]
```diff
@@ -49,3 +76,21 @@ const getOffset = (el) => {
...
   return { top: rect.top + scrollTop, left: rect.left + scrollLeft };
 };
+
+const validatePosition = (position, PositionCheck, ShiftHandler) => {
+  if (PositionCheck[position]()) {
+    return { ...ShiftHandler[position], newPosition: position };
+  }
+
+  for (const key in Position) {
+    const side = Position[key];
```
Comment (v-excelsior) [inline components/bl-popover-component/src/helpers.js:86]: Object.values(Position).forEach

File: components/bl-popover-component/src/helpers.js [hunk 4]
```diff
@@ -1,45 +1,72 @@
...
     },
     [Position.BOTTOM]: {
-      topShift : contentOffset.top + contentHeight - rootOffset.top + root.firstChild.clientHeight,
+      topShift : targetOffset.top + targetHeight - contentOffset.top + contentElement.firstChild.clientHeight,
       leftShift: verticalLeftShift
     }
   };
 
-  return ShiftHandler[position];
+  const PositionCheck = {
```
Comment (v-excelsior) [inline components/bl-popover-component/src/helpers.js:46]: PositionCheck - PositionValidator

File: components/bl-popover-component/src/helpers.js [hunk 5]
```diff
@@ -1,45 +1,72 @@
...
+        && (targetY + targetHorizontalCenter) >= contentHorizontalCenter
+        && window.innerHeight - targetBottom + targetHorizontalCenter >= contentHorizontalCenter
+    ),
+    top   : () => (
+      targetY >= contentHeight
+        && (targetX + targetVerticalCenter) >= contentVerticalCenter
+        && document.documentElement.clientWidth - targetRight + targetVerticalCenter >= contentVerticalCenter
+    ),
+    bottom: () => (
+      window.innerHeight - targetBottom >= contentHeight
```
Comment (v-excelsior) [inline components/bl-popover-component/src/helpers.js:63]: I'm not sure if it good approach, and my offer is to do it(check for intersection) through IntersectionObserver, but I'm not insisting.

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-popover-component/src/helpers.js (hunk 3 above)
Comment (StasivPavlo) [reply components/bl-popover-component/src/helpers.js:86]: but return doesn't work inside forEach

**Pull request merged**